### DISCUSSION TITLE
UiXml previewer problems

### DISCUSSION THREAD

**Discussion by generalloki** [alternetsoft/AlternetUI#236] [Ideas]
UiXml previewer has some problems:

I collected them from the #15:

- Referenced images are not correctly loaded by UiXml previewer. Previewer doesn't like "embres:MenuSample.Resources.Icons.Small.Photo16.png"
- Local controls are not corretly loaded by UiXml previewer. Previewer doesn't like <local:SomeControl /> when they have own xml.
- App need to be compiled before UiXml previewer start to work.
- Window events are called UiXml previewer but in the strange maneer (not the same way as it is usually done by the application during normal start). This causes different exceptions.
- WebBrowser preview not working
- Errors in preview not easy to understand
- If uiXML is not valid - the previewer does not display anything.
- UiXML code completion should not display enumerations as tags.

**Comment by generalloki** [comment]
Also preview of Menu or Toolbar shows Windows desktop

**Comment by neoxeo** [comment]
Samples/ControlsTest : WebBrowserTestPage.uixml have error (de4a2816ad871122cc606ee0dfed220a7c465b0e)
```
13:58:04.886 [Verbose] 10540 Started PreviewerProcess.OnMessageAsync()
13:58:04.886 [Debug] 10540 <= UpdateXamlResultMessage { Error: "Error while updating UIXML prev…", Handle: null, Exception: ExceptionDetails { ExceptionType: "TypeInitializationException", Message: "The type initializer for 'Contr…", StackTrace: "   at System.RuntimeMethodHandl…", UixmlLineNumber: null, UixmlLinePosition: null } }
13:58:04.886 [Error] 10540 UpdateXamlResult error
System.Xaml.XamlException: The type initializer for 'ControlsTest.WebBrowserTestPage' threw an exception.
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.ConstructorInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
13:58:04.886 [Verbose] 10540 Finished PreviewerProcess.OnMessageAsync()
```

**Comment by neoxeo** [comment]
Samples/PaintSample : MainWindow.uixml have error (https://github.com/alternetsoft/AlternetUI/commit/de4a2816ad871122cc606ee0dfed220a7c465b0e)
```
14:21:58.873 [Verbose]  Finished PreviewerProcess.Stop()
14:21:58.858 [Error]  Connection error
System.IO.IOException: Impossible de lire les données de la connexion de transport : Une connexion existante a dû être fermée par l’hôte distant. ---> System.Net.Sockets.SocketException: Une connexion existante a dû être fermée par l’hôte distant
   à System.Net.Sockets.Socket.EndReceive(IAsyncResult asyncResult)
   à System.Net.Sockets.NetworkStream.EndRead(IAsyncResult asyncResult)
   --- Fin de la trace de la pile d'exception interne ---
   à System.Net.Sockets.NetworkStream.EndRead(IAsyncResult asyncResult)
   à System.Threading.Tasks.TaskFactory`1.FromAsyncTrimPromise`1.Complete(TInstance thisRef, Func`3 endMethod, IAsyncResult asyncResult, Boolean requiresSynchronization)
--- Fin de la trace de la pile à partir de l'emplacement précédent au niveau duquel l'exception a été levée ---
   à System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   à Alternet.UI.Integration.Remoting.BsonStreamTransportConnection.<ReadExact>d__16.MoveNext()
--- Fin de la trace de la pile à partir de l'emplacement précédent au niveau duquel l'exception a été levée ---
   à System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   à Alternet.UI.Integration.Remoting.BsonStreamTransportConnection.<Reader>d__17.MoveNext()
```

**Comment by generalloki** [comment]
I have uploaded 0.9.620-beta and it has an updated uixml preview which works ok.
https://github.com/alternetsoft/AlternetUI/releases

There are still some problems:

- We need to wait for some time when Visual Studio started before opening uixml or we have an exception.
- We need to build project before opening uixml
- Uixml from ControlsSampleDll and maybe other dlls are not opening, only uixml from exe projects are opened.

**Comment by generalloki** [comment]
In the 0.9.759 build we improved uixml previewer. Now it doesn't raise an exception when file is not ready, it shows message about it.

**Comment by generalloki** [comment]
For now I am converting this to discussion. After 1.0 release we will look into this
